File: java/src/main/java/com/example/s3service/controller/S3Controller.java
package com.example.s3service.controller;

import com.example.s3service.service.S3Service;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import software.amazon.awssdk.core.ResponseBytes;
import software.amazon.awssdk.services.s3.model.GetObjectResponse;

/**
 * S3 REST Controller
 * 
 * @author KO
 * @since 2026-01-19
 */
@RestController
@RequestMapping("/api/s3")
public class S3Controller {

    private final S3Service s3Service;

    /**
     * Constructor with S3Service dependency injection
     * 
     * @param s3Service S3Service instance
     */
    public S3Controller(S3Service s3Service) {
        this.s3Service = s3Service;
    }

    /**
     * Upload file to S3 bucket
     * 
     * @param file MultipartFile to upload
     * @param key  optional S3 object key, if not provided, use original filename
     * @param bucket optional bucket name, if not provided, use default bucket
     * @return ResponseEntity with upload result
     */
    @PostMapping("/upload")
    public ResponseEntity<String> uploadFile(@RequestParam("file") MultipartFile file, 
                                             @RequestParam(value = "key", required = false) String key, 
                                             @RequestParam(value = "bucket", required = false) String bucket) {
        try {
            String objectKey = key != null && !key.isEmpty() ? key : file.getOriginalFilename();
            s3Service.uploadFile(file, objectKey, bucket);
            return ResponseEntity.ok("File uploaded successfully with key: " + objectKey);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body("Failed to upload file: " + e.getMessage());
        }
    }

    /**
     * Download file from S3 bucket
     * 
     * @param key S3 object key
     * @param bucket optional bucket name, if not provided, use default bucket
     * @return ResponseEntity with file content
     */
    @GetMapping("/download/{key}")
    public ResponseEntity<ByteArrayResource> downloadFile(@PathVariable String key, 
                                                          @RequestParam(value = "bucket", required = false) String bucket) {
        try {
            ResponseBytes<GetObjectResponse> responseBytes = s3Service.downloadFile(key, bucket);
            byte[] data = responseBytes.asByteArray();
            GetObjectResponse response = responseBytes.response();

            ByteArrayResource resource = new ByteArrayResource(data);

            return ResponseEntity.ok()
                    .contentType(MediaType.parseMediaType(response.contentType()))
                    .contentLength(data.length)
                    .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + key + "\"")
                    .body(resource);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(null);
        }
    }

    /**
     * Delete file from S3 bucket
     * 
     * @param key S3 object key
     * @param bucket optional bucket name, if not provided, use default bucket
     * @return ResponseEntity with delete result
     */
    @DeleteMapping("/delete/{key}")
    public ResponseEntity<String> deleteFile(@PathVariable String key, 
                                             @RequestParam(value = "bucket", required = false) String bucket) {
        try {
            s3Service.deleteFile(key, bucket);
            return ResponseEntity.ok("File deleted successfully: " + key);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body("Failed to delete file: " + e.getMessage());
        }
    }

    /**
     * Check if file exists in S3 bucket
     * 
     * @param key S3 object key
     * @param bucket optional bucket name, if not provided, use default bucket
     * @return ResponseEntity with exists status
     */
    @GetMapping("/exists/{key}")
    public ResponseEntity<Boolean> checkFileExists(@PathVariable String key, 
                                                  @RequestParam(value = "bucket", required = false) String bucket) {
        try {
            boolean exists = s3Service.fileExists(key, bucket);
            return ResponseEntity.ok(exists);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(false);
        }
    }

    /**
     * Health check endpoint
     * 
     * @return ResponseEntity with health status
     */
    @GetMapping("/health")
    public ResponseEntity<String> healthCheck() {
        return ResponseEntity.ok("S3 Service is running");
    }

    /**
     * List all files in S3 bucket
     * 
     * @param bucket optional bucket name, if not provided, use default bucket
     * @return ResponseEntity with list of files
     */
    @GetMapping("/list")
    public ResponseEntity<?> listFiles(@RequestParam(value = "bucket", required = false) String bucket) {
        try {
            var response = s3Service.listFiles(bucket);
            // 转换为前端可以处理的简单数据结构
            var files = response.contents().stream()
                    .map(file -> {
                        // 创建包含必要信息的简单对象
                        var fileInfo = new java.util.HashMap<String, Object>();
                        fileInfo.put("key", file.key());
                        fileInfo.put("size", file.size());
                        fileInfo.put("lastModified", file.lastModified());
                        return fileInfo;
                    })
                    .collect(java.util.stream.Collectors.toList());
            return ResponseEntity.ok(files);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body("Failed to list files: " + e.getMessage());
        }
    }
    
    /**
     * List all S3 buckets
     * 
     * @return ResponseEntity with list of buckets
     */
    @GetMapping("/buckets")
    public ResponseEntity<?> listBuckets() {
        try {
            var buckets = s3Service.listBuckets();
            // 转换为前端可以处理的简单数据结构
            var bucketInfo = buckets.stream()
                    .map(bucket -> {
                        // 创建包含必要信息的简单对象
                        var info = new java.util.HashMap<String, Object>();
                        info.put("name", bucket.name());
                        info.put("creationDate", bucket.creationDate());
                        return info;
                    })
                    .collect(java.util.stream.Collectors.toList());
            return ResponseEntity.ok(bucketInfo);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body("Failed to list buckets: " + e.getMessage());
        }
    }
    
    /**
     * Create a new S3 bucket
     * 
     * @param bucketName name of the bucket to create
     * @return ResponseEntity with creation result
     */
    @PostMapping("/bucket")
    public ResponseEntity<String> createBucket(@RequestParam("bucketName") String bucketName) {
        try {
            boolean created = s3Service.createBucket(bucketName);
            if (created) {
                return ResponseEntity.ok("Bucket created successfully: " + bucketName);
            } else {
                return ResponseEntity.status(HttpStatus.CONFLICT)
                        .body("Bucket already exists: " + bucketName);
            }
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body("Failed to create bucket: " + e.getMessage());
        }
    }
}

File: go/controllers/s3_controller.go
// Package controllers 处理HTTP请求
// 作者: KO
// 创建时间: 2026-01-27
// 修改时间: 2026-01-27
package controllers

import (
	"bytes"
	"fmt"
	"net/http"

	"github.com/example/s3service/s3"
	"github.com/labstack/echo/v4"
)

// S3Controller 处理S3相关的HTTP请求
type S3Controller struct {
	service *s3.Service // S3服务实例
}

// NewS3Controller 创建新的S3控制器实例
// 参数:
//
//	service: S3服务实例
//
// 返回值:
//
//	*S3Controller: S3控制器实例
func NewS3Controller(service *s3.Service) *S3Controller {
	return &S3Controller{
		service: service,
	}
}

// UploadFile 上传文件到S3存储桶
// 参数:
//
//	ctx: Echo上下文
//
// 返回值:
//
//	error: 错误信息
func (c *S3Controller) UploadFile(ctx echo.Context) error {
	file, err := ctx.FormFile("file")
	if err != nil {
		return ctx.JSON(http.StatusBadRequest, map[string]string{
			"error": "Invalid file",
		})
	}

	// 打开上传的文件
	src, err := file.Open()
	if err != nil {
		return ctx.JSON(http.StatusInternalServerError, map[string]string{
			"error": "Failed to open file",
		})
	}
	defer src.Close()

	// 读取文件内容
	content := bytes.Buffer{}
	if _, err := content.ReadFrom(src); err != nil {
		return ctx.JSON(http.StatusInternalServerError, map[string]string{
			"error": "Failed to read file",
		})
	}

	// 获取对象键
	key := ctx.FormValue("key")
	if key == "" {
		key = file.Filename
	}

	// 获取存储桶名称
	bucket := ctx.FormValue("bucket")

	// 上传文件
	if err := c.service.UploadFile(ctx.Request().Context(), bucket, key, content.Bytes()); err != nil {
		return ctx.JSON(http.StatusInternalServerError, map[string]string{
			"error": "Failed to upload file: " + err.Error(),
		})
	}

	return ctx.JSON(http.StatusOK, map[string]string{
		"message": "File uploaded successfully with key: " + key,
	})
}

// DownloadFile 从S3存储桶下载文件
// 参数:
//
//	ctx: Echo上下文
//
// 返回值:
//
//	error: 错误信息
func (c *S3Controller) DownloadFile(ctx echo.Context) error {
	key := ctx.Param("key")
	bucket := ctx.QueryParam("bucket")

	content, err := c.service.DownloadFile(ctx.Request().Context(), bucket, key)
	if err != nil {
		return ctx.JSON(http.StatusInternalServerError, map[string]string{
			"error": "Failed to download file: " + err.Error(),
		})
	}

	// 设置响应头
	ctx.Response().Header().Set("Content-Disposition", "attachment; filename="+key)
	ctx.Response().Header().Set("Content-Length", fmt.Sprintf("%d", len(content)))

	return ctx.Blob(http.StatusOK, "application/octet-stream", content)
}

// DeleteFile 从S3存储桶删除文件
// 参数:
//
//	ctx: Echo上下文
//
// 返回值:
//
//	error: 错误信息
func (c *S3Controller) DeleteFile(ctx echo.Context) error {
	key := ctx.Param("key")
	bucket := ctx.QueryParam("bucket")

	if err := c.service.DeleteFile(ctx.Request().Context(), bucket, key); err != nil {
		return ctx.JSON(http.StatusInternalServerError, map[string]string{
			"error": "Failed to delete file: " + err.Error(),
		})
	}

	return ctx.JSON(http.StatusOK, map[string]string{
		"message": "File deleted successfully: " + key,
	})
}

// CheckFileExists 检查文件是否存在于S3存储桶
// 参数:
//
//	ctx: Echo上下文
//
// 返回值:
//
//	error: 错误信息
func (c *S3Controller) CheckFileExists(ctx echo.Context) error {
	key := ctx.Param("key")
	bucket := ctx.QueryParam("bucket")

	exists := c.service.FileExists(ctx.Request().Context(), bucket, key)

	return ctx.JSON(http.StatusOK, exists)
}

// HealthCheck 健康检查端点
// 参数:
//
//	ctx: Echo上下文
//
// 返回值:
//
//	error: 错误信息
func (c *S3Controller) HealthCheck(ctx echo.Context) error {
	return ctx.String(http.StatusOK, "S3 Service is running")
}

// ListFiles 列出S3存储桶中的所有文件
// 参数:
//
//	ctx: Echo上下文
//
// 返回值:
//
//	error: 错误信息
func (c *S3Controller) ListFiles(ctx echo.Context) error {
	bucket := ctx.QueryParam("bucket")

	files, err := c.service.ListFiles(ctx.Request().Context(), bucket)
	if err != nil {
		return ctx.JSON(http.StatusInternalServerError, map[string]string{
			"error": "Failed to list files: " + err.Error(),
		})
	}

	return ctx.JSON(http.StatusOK, files)
}

// ListBuckets 列出所有S3存储桶
// 参数:
//
//	ctx: Echo上下文
//
// 返回值:
//
//	error: 错误信息
func (c *S3Controller) ListBuckets(ctx echo.Context) error {
	buckets, err := c.service.ListBuckets(ctx.Request().Context())
	if err != nil {
		return ctx.JSON(http.StatusInternalServerError, map[string]string{
			"error": "Failed to list buckets: " + err.Error(),
		})
	}

	return ctx.JSON(http.StatusOK, buckets)
}

// CreateBucket 创建新的S3存储桶
// 参数:
//
//	ctx: Echo上下文
//
// 返回值:
//
//	error: 错误信息
func (c *S3Controller) CreateBucket(ctx echo.Context) error {
	bucketName := ctx.QueryParam("bucketName")
	if bucketName == "" {
		return ctx.JSON(http.StatusBadRequest, map[string]string{
			"error": "Bucket name is required",
		})
	}

	if err := c.service.CreateBucket(ctx.Request().Context(), bucketName); err != nil {
		if err.Error() == "bucket already exists" {
			return ctx.JSON(http.StatusConflict, map[string]string{
				"error": "Bucket already exists: " + bucketName,
			})
		}
		return ctx.JSON(http.StatusInternalServerError, map[string]string{
			"error": "Failed to create bucket: " + err.Error(),
		})
	}

	return ctx.JSON(http.StatusOK, map[string]string{
		"message": "Bucket created successfully: " + bucketName,
	})
}